### ISSUE TITLE
Allow to reconnect to running instance

### ISSUE THREAD

**Issue by annappropriate** [vorburger/MariaDB4j#98]
I'm currently developing a test suite for an app, and I'm facing the following problem: there's no way to keep an instance of `DB` object around between tests, and I do need it around.

What I would like to do is 
```
val db: DB =
    if (isDbRunning) {
        DB.reconnect(port)
     } else {
        val newDb = DB.newEmbeddedDB(port)
        newDb.start()
        newDb
     }
```
(my implementation  of `isDbRunning` is based on 
```
  private def isPortInUse(port: Int): Boolean = try {
    (new ServerSocket(port)).close()
    false
  } catch {
    case _: SocketException => true
  }
```
If this ends up in library, it could be an entirely contained method like `DB.createOrReconnect(port|config)`

The problem with this idea is that `DB` constructor is protected. I tried creating a successor class and exposing a build method, but then `baseDir` in `DB` is not initialized. If I call `DB#prepareDirectories`, it destroys the `dataDir`. Overriding `prepareDirectories` is impossible because `baseDir`, `libDir` and `dataDir` are `private`.

Apparently destruction of `dataDir` leads to this exception:
```
[info]   Cause: ch.vorburger.exec.ManagedProcessException: Program [/tmp/MariaDB4j/base/bin/mysql, --socket=/tmp/MariaDB4j.3310.sock] (in working directory /tmp/MariaDB4j/base) failed, exitValue=1, last 100 lines of console:
[info] ERROR 1006 (HY000) at line 1: Can't create database 'foo' (errno: 2 "No such file or directory")
[info]   at ch.vorburger.exec.ManagedProcess.checkResult(ManagedProcess.java:297)
[info]   at ch.vorburger.exec.ManagedProcess.startExecute(ManagedProcess.java:207)
[info]   at ch.vorburger.exec.ManagedProcess.start(ManagedProcess.java:134)
[info]   at ch.vorburger.mariadb4j.DB.run(DB.java:267)
[info]   at ch.vorburger.mariadb4j.DB.source(DB.java:238)
[info]   at ch.vorburger.mariadb4j.DB.source(DB.java:221)
...
```

What do you think about this? I could spend some time to prepare a PR if you think this idea is worthy. Restarting database between every testsuite takes quite a long time :(

**Comment by vorburger** [comment]
@annappropriate thanks for taking the time to open this issue - appreciate your contribution to MariaDB4j.

But I have to admit I'm not following what you want to do, how do you want to reconnect? I don't see how a `java.lang.Process` can be re-attached to a PID... do you?

Why can you not keep an instance of DB object around between tests? That's not a MariaDB4j problem, but an issue with your favourite test framework.. I suspect you are using JUnit, and because it create new instances for each of your test classes, you don't know where to keep the DB instance around? What you could do is to just write something like a `class SingletonDB { public static DB db; }` ...

Re. destroys the dataDir, please see e.g. https://github.com/vorburger/MariaDB4j/issues/94

**Comment by annappropriate** [comment]
You're right, I also don't know how to reattach to PID.

I'm using scalatest. It does drop `object`s between test suites. I know that I can create a test suite and add test classes there manually, but I find the implementation ugly and was looking for alternative approach.

After some thinking I figured that my confusion about the issue I raised stems from the fact that `DB` acts both as server(`start`/`stop`) and client (`source`/`run`) from my perspective, and after calling `start` once on a server, I only need the client functionality. That is the reason I was trying to rebuild `DB` and stuff. Should I be better off using some other client then?
